src/ecephys_etl/modules/align_timestamps/barcode.py

```diff
@@ -0,0 +1,303 @@
+    barcode_duration_ceiling : numeric, optional 
+        The maximum duration of a single barcode
+    nbits : int, optional
+        The bit-depth of each barcode
+
+    Returns
+    -------
+    barcode_start_times : list of numeric
+        For each detected barcode, the time at which that barcode started
+    barcodes : list of int
+        For each detected barcode, the value of that barcode as an integer.
+
+    Notes
+    -----
+    ignores first code in prod (ok, but not intended)
+    ignores first on pulse (intended - this is needed to identify that a barcode is starting)
+
+    """
+
+    start_indices = np.diff(on_times)
+    a = np.where(start_indices > inter_barcode_interval)[0]
+    barcode_start_times = on_times[a + 1]
+
+    barcodes = []
+
+    for i, t in enumerate(barcode_start_times):
+
+        oncode = on_times[
+            np.where(
+                np.logical_and(on_times > t, on_times < t + barcode_duration_ceiling)
+            )[0]
+        ]
+        offcode = off_times[
+            np.where(
+                np.logical_and(off_times > t, off_times < t + barcode_duration_ceiling)
+            )[0]
+        ]
+
+        currTime = offcode[0]
+
+        bits = np.zeros((nbits,))
+
+        for bit in range(0, nbits):
+
+            nextOn = np.where(oncode > currTime)[0]
+            nextOff = np.where(offcode > currTime)[0]
+
+            if nextOn.size > 0:
+                nextOn = oncode[nextOn[0]]
+            else:
+                nextOn = t + inter_barcode_interval
+
+            if nextOff.size > 0:
+                nextOff = offcode[nextOff[0]]
+            else:
+                nextOff = t + inter_barcode_interval
+
+            if nextOn < nextOff:
+                bits[bit] = 1
+
+            currTime += bar_duration
+
+        barcode = 0
+
+        # least sig left
+        for bit in range(0, nbits):
+            barcode += bits[bit] * pow(2, bit)
+
+        barcodes.append(barcode)
+
+    return barcode_start_times, barcodes
+
+
+def find_matching_index(master_barcodes, probe_barcodes, alignment_type="start"):
+    """Given a set of barcodes for the master clock and the probe clock, find the
+    indices of a matching set, either starting from the beginning or the end
+    of the list.
+
+    Parameters
+    ----------
+    master_barcodes : np.ndarray
+        barcode values on the master line. One per barcode
+    probe_barcodes : np.ndarray
+        barcode values on the probe line. One per barcode
+    alignment_type : string
+        'start' or 'end'
+
+    Returns
+    -------
+    master_barcode_index : int
+        matching index for master barcodes (None if not found)
+    probe_barcode_index : int
+        matching index for probe barcodes (None if not found)
+
+    """
+
+    foundMatch = False
+    master_barcode_index = None
+
+    if alignment_type == "start":
+        probe_barcode_index = 0
+        direction = 1
+    else:
```

Comment:
Should this be `elif alignment_type == "end":` followed by an `else` block that raises an exception if an alignment type other than `"start"` or `"end"` is passed in?